server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java

```diff
@@ -247,32 +248,29 @@ private boolean cancelTaskIfGranularityChanged(
 
   /**
    * Gets a List of Intervals locked by higher priority tasks for each datasource.
+   * However, when using a REPLACE lock for compaction, intervals locked with any APPEND lock will not be returned
    * Since compaction tasks submitted for these Intervals would have to wait anyway,
    * we skip these Intervals until the next compaction run.
    * <p>
    * For now, Segment Locks are being treated the same as Time Chunk Locks even
    * though they lock only a Segment and not the entire Interval. Thus,
    * a compaction task will not be submitted for an Interval if
    * <ul>
-   *   <li>either the whole Interval is locked by a higher priority Task</li>
+   *   <li>either the whole Interval is locked by a higher priority Task with an incompatible lock type</li>
    *   <li>or there is atleast one Segment in the Interval that is locked by a
    *   higher priority Task</li>
    * </ul>
    */
-  private Map<String, List<Interval>> getLockedIntervalsToSkip(
+  private Map<String, List<Interval>> getLockedIntervalsV2(
       List<DataSourceCompactionConfig> compactionConfigs
   )
   {
-    final Map<String, Integer> minTaskPriority = compactionConfigs
+    final List<LockFilterPolicy> lockFilterPolicies = compactionConfigs
         .stream()
-        .collect(
-            Collectors.toMap(
-                DataSourceCompactionConfig::getDataSource,
-                DataSourceCompactionConfig::getTaskPriority
-            )
-        );
+        .map(config -> new LockFilterPolicy(config.getDataSource(), config.getTaskPriority(), config.getTaskContext()))
```

Comment:
Why are we sending the whole task context over the API? Why not just the lock type?

Reply:
Tasks.TASK_LOCK_TYPE, TaskLockType etc are not available here.
Didn't want to hardcode the string passed as the key

Reply:
The names of these context variables are not going to change, so hardcoding is not that big of a concern. Having to pass the whole context is probably not great. That makes this API very specific to auto-compaction duty.